src/middlewared/middlewared/plugins/reporting.py

```diff
@@ -655,14 +666,201 @@ class NFSStatPlugin(RRDBase):
+
+        await self.middleware.call('service.restart', 'collectd')
+
+        return await self.config()
+
+    @private
+    def setup(self):
+        is_freenas = self.middleware.call_sync('system.is_freenas')
+        # If not is_freenas remove the rc.conf cache rc.conf.local will
+        # run again using the correct collectd_enable. See #5019
+        if not is_freenas:
+            try:
+                os.remove('/var/tmp/freenas_config.md5')
+            except FileNotFoundError:
+                pass
+
+        systemdatasetconfig = self.middleware.call_sync('systemdataset.config')
+        if not systemdatasetconfig['path']:
+            self.middleware.logger.error(f'System dataset is not mounted')
+            return False
+
+        rrd_mount = f'{systemdatasetconfig["path"]}/rrd-{systemdatasetconfig["uuid"]}'
+        if not os.path.exists(rrd_mount):
+            self.middleware.logger.error(f'{rrd_mount} does not exist or is not a directory')
+            return False
+
+        # Ensure that collectd working path is a symlink to system dataset
+        pwd = '/var/db/collectd/rrd'
+        if os.path.exists(pwd) and (not os.path.isdir(pwd) or not os.path.islink(pwd)):
+            shutil.move(pwd, f'{pwd}.{time.strftime("%Y%m%d%H%M%S")}')
+        if not os.path.exists(pwd):
+            os.symlink(rrd_mount, pwd)
+
+        # Migrate legacy RAMDisk
+        persist_file = '/data/rrd_dir.tar.bz2'
+        if os.path.isfile(persist_file):
+            with tarfile.open(persist_file) as tar:
+                if 'collectd/rrd' in tar.getnames():
+                    tar.extractall(pwd, get_members(tar, 'collectd/rrd/'))
+
+            os.unlink('/data/rrd_dir.tar.bz2')
+
+        hostname = self.middleware.call_sync('system.info')['hostname']
+        if not hostname:
+            hostname = self.middleware.call_sync('network.configuration.config')['hostname']
+
+        # Migrate from old version, where `hostname` was a real directory and `localhost` was a symlink.
+        # Skip the case where `hostname` is "localhost", so symlink was not (and is not) needed.
+        if (
+            hostname != 'localhost' and
+            os.path.isdir(os.path.join(pwd, hostname)) and
+            not os.path.islink(os.path.join(pwd, hostname))
+        ):
+            if os.path.exists(os.path.join(pwd, 'localhost')):
+                if os.path.islink(os.path.join(pwd, 'localhost')):
+                    os.unlink(os.path.join(pwd, 'localhost'))
+                else:
+                    # This should not happen, but just in case
+                    shutil.move(
+                        os.path.join(pwd, 'localhost'),
+                        os.path.join(pwd, f'localhost.bak.{time.strftime("%Y%m%d%H%M%S")}')
+                    )
+            shutil.move(os.path.join(pwd, hostname), os.path.join(pwd, 'localhost'))
+
+        # Remove all directories except "localhost" and it's backups (that may be erroneously created by
```

Comment:
`` # Remove all directories except "localhost" and its backups (that may be erroneously created by ``

(Possessive version of "it" does not have an apostrophe.  Another reason to avoid possessives entirely.)